### PR TITLE
(refactor): Enable new engines with custom dispatching and other constructs

### PR DESCRIPTION
**Pull request by jaidisido**
### Feature or Bugfix

- Refactoring

### Detail
The current codebase cannot easily support multiple engines. The below changes remediate that:
- Move distributed methods out of non-distributed modules
- Add a custom dispatcher and use it to register methods based on the execution engine and memory format value
- Add classes to manage the state of the execution engine and memory format
- Refactor the structure of distributed modules to support multiple engines

Code flow:
1. Engine [class](https://github.com/aws/aws-sdk-pandas/blob/refactor-dispatching/awswrangler/_distributed.py#L17) is initialised at [import](https://github.com/aws/aws-sdk-pandas/blob/refactor-dispatching/awswrangler/__init__.py#L37)
2. It detects the installed engine (python vs ray), [initialises](https://github.com/aws/aws-sdk-pandas/blob/refactor-dispatching/awswrangler/_distributed.py#L73) the engine and [registers](https://github.com/aws/aws-sdk-pandas/blob/refactor-dispatching/awswrangler/distributed/ray/_register.py) the dispatched methods based on the value of the engine and memory format (using this custom [dispatcher](https://github.com/aws/aws-sdk-pandas/blob/refactor-dispatching/awswrangler/_dispatch.py))

The good:
- Clear separation of concerns: Distributed methods live outside non-distributed code, eliminating ugly if conditionals, allowing both to scale independently and making them easier to maintain in the future
- Better dispatching: Adding a new engine/memory format is as simple as creating a new directory with its methods and registering them with the custom dispatcher based on the value of the engine or memory format
- Custom engine/memory format classes: Give more flexibility than `config` when it comes to interacting with the engine and managing its state (initialising, registering, get/setting...)

The bad:
- Managing state: Adding a custom dispatcher means that we must maintain its state. For instance, one capability currently missing in this PR is **unregistering** methods when a user sets a different engine (e.g. moving from ray to dask at execution time). 
- Detecting the engine: Conditionals are simpler/easier when it comes to detecting an engine. With a custom dispatcher, the registration and dispatching process is more opaque/convoluted. For instance, there is a higher risk of not realising that we are using a given engine vs another

The ugly:
- Unused arguments: Each method registered with the dispatcher must accept the union of both non-distributed and distributed arguments, even though some would be unused. As the list of supported engines grows, so does the number of unused arguments. It also means that we must maintain the same list of arguments across the different versions of the method

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by malachi-constant** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubDistributedCodeBuild6-jWcl5DLmvupS
* Commit ID: 7e5b8e30184137902665f17107c8e9eb58668327
* Result: SUCCEEDED
* [Build Logs](https://p0rshq9q5i.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=ff798235-fec9-4806-ace2-29f47d3666f4%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by malachi-constant** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubLoadTests5656BB24-s6u9F3qN9oFy
* Commit ID: 7e5b8e30184137902665f17107c8e9eb58668327
* Result: SUCCEEDED
* [Build Logs](https://pciyom2pbh.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=ba2dd23c-2cb1-4eb4-af6b-db56bfec5336%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by malachi-constant** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubStandardCodeBuild8C06-llutOAimTATs
* Commit ID: 7e5b8e30184137902665f17107c8e9eb58668327
* Result: SUCCEEDED
* [Build Logs](https://jsv9u8gote.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=7e07585d-a2b6-4600-8e30-46e2a7ddd785%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

File: awswrangler/s3/_write_parquet.py [hunk 1]
```diff
@@ -224,54 +211,6 @@ def _to_parquet(
     return paths
 
 
-def _to_parquet_distributed(  # pylint: disable=unused-argument
```
Comment (kukushking) [inline awswrangler/s3/_write_parquet.py:227]: Thanks for moving these out into a separate package! Less conditional imports, no more mixed distributed and non-distributed code. Definitely a highlight of this PR!

File: awswrangler/distributed/ray/_register.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+from awswrangler.s3._write_text import _to_text
+
+
+def register_ray() -> None:
+    """Register dispatched Ray and Modin (on Ray) methods."""
+    execution_engine = EngineEnum.RAY.value
+
+    _batch_paths.register(execution_engine, _batch_paths_distributed)  # type: ignore
+
+    if memory_format.get() == MemoryFormatEnum.MODIN.value:
```
Comment (LeonLuttenberger) [inline awswrangler/distributed/ray/_register.py:20]: Really cool, I love how this is out of way!

**Review by LeonLuttenberger — APPROVED** [review]
Awesome work!

**Comment by malachi-constant** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubDistributedCodeBuild6-jWcl5DLmvupS
* Commit ID: 76ac88e10f28e5ec3b47d09c5186a3e2eff5bde9
* Result: FAILED
* [Build Logs](https://p0rshq9q5i.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=9e36dbfa-fd34-4251-8dc6-df8e0c834839%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by malachi-constant** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubLoadTests5656BB24-s6u9F3qN9oFy
* Commit ID: 76ac88e10f28e5ec3b47d09c5186a3e2eff5bde9
* Result: FAILED
* [Build Logs](https://pciyom2pbh.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=36fb936c-5cf1-4a1d-abbe-ce128943e3f9%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by malachi-constant** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubStandardCodeBuild8C06-llutOAimTATs
* Commit ID: 76ac88e10f28e5ec3b47d09c5186a3e2eff5bde9
* Result: FAILED
* [Build Logs](https://jsv9u8gote.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=6bec3a8b-011e-425e-a0a7-c4b8d4120541%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

File: awswrangler/_distributed.py [hunk 3]
```diff
@@ -0,0 +1,152 @@
+"""Distributed engine and memory format configuration."""
```
Comment (malachi-constant) [inline awswrangler/_distributed.py:1]: Nice, handling this logic here seems a lot cleaner to me, great work :)

**Comment by malachi-constant** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubDistributedCodeBuild6-jWcl5DLmvupS
* Commit ID: c511d30dd94f0bf068126ccdb84dae7dd0178450
* Result: SUCCEEDED
* [Build Logs](https://p0rshq9q5i.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=1fc663d8-2e3a-4260-81cc-8f7db51d2a47%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by malachi-constant** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubStandardCodeBuild8C06-llutOAimTATs
* Commit ID: c511d30dd94f0bf068126ccdb84dae7dd0178450
* Result: SUCCEEDED
* [Build Logs](https://jsv9u8gote.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=f1b373a2-d366-4759-9fe0-9e4ecf14b494%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by malachi-constant** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubLoadTests5656BB24-s6u9F3qN9oFy
* Commit ID: c511d30dd94f0bf068126ccdb84dae7dd0178450
* Result: SUCCEEDED
* [Build Logs](https://pciyom2pbh.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=06af27fe-d389-4ac5-9954-3bfb215d0338%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by malachi-constant** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubDistributedCodeBuild6-jWcl5DLmvupS
* Commit ID: 9fd4093a617af60ccca4ad3a212987cb1acaec1c
* Result: SUCCEEDED
* [Build Logs](https://p0rshq9q5i.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=485dc4a8-ddb7-416b-8284-34a3e3bc4e42%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by malachi-constant** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubLoadTests5656BB24-s6u9F3qN9oFy
* Commit ID: 9fd4093a617af60ccca4ad3a212987cb1acaec1c
* Result: SUCCEEDED
* [Build Logs](https://pciyom2pbh.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=09e4f8a0-1b56-4ef3-9191-1a25be46bbdf%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by malachi-constant** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubStandardCodeBuild8C06-llutOAimTATs
* Commit ID: 9fd4093a617af60ccca4ad3a212987cb1acaec1c
* Result: SUCCEEDED
* [Build Logs](https://jsv9u8gote.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=30fdfa89-51b5-4313-9597-51ff1c6fb1ac%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

File: awswrangler/distributed/ray/_register.py (hunk 2 above)
Comment (kukushking) [reply awswrangler/distributed/ray/_register.py:29]: +1. Minor Q - can `memory_format.get()` return an enum so you don't have to do `.value` on every comparison?

File: awswrangler/distributed/ray/_register.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    """Register dispatched Ray and Modin (on Ray) methods."""
+    # LakeFormation
+    engine.register_func(_get_work_unit_results, ray_remote(_get_work_unit_results))
+
+    # S3
+    engine.register_func(_delete_objects, ray_remote(_delete_objects))
+    engine.register_func(_read_parquet_metadata_file, ray_remote(_read_parquet_metadata_file))
+    engine.register_func(_select_query, ray_remote(_select_query))
+    engine.register_func(_select_object_content, ray_remote(_select_object_content))
+    engine.register_func(_wait_object_batch, ray_remote(_wait_object_batch))
```
Comment (kukushking) [inline awswrangler/distributed/ray/_register.py:27]: What do you think about keeping the register only for registration and moving the function definitions to modin module with the others? E.g.
```
distributed/ray/modin/s3/...py:

_wait_object_batch_distributed = ray_remote(_wait_object_batch)
_select_object_content_distributed = ray_remote(_select_object_content)
```
Also it seems now we do not actually need our own `ray_remote` decorator that handled both distributed and non-distributed calls dynamically and we can use default `@remote`.

File: awswrangler/_distributed.py [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+        return wrapper
+
+    @classmethod
+    def register(cls, name: Optional[str] = None) -> None:
+        """Register the distribution engine dispatch methods."""
+        engine_name = cls.get_installed() if not name else name
+        cls.set(engine_name)
+        cls._registry.clear()
+
+        if engine_name == EngineEnum.RAY.value:
```
Comment (kukushking) [inline awswrangler/_distributed.py:98]: Looks great! In the future we might want to add concrete implementations of Engine e.g. `RayEngine` etc that adhere to common Engine interface but contain all engine-specific logic e.g. setup/teardown, register.

**Review by kukushking — APPROVED** [review]
Awesome!

**Comment by malachi-constant** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubStandardCodeBuild8C06-llutOAimTATs
* Commit ID: fadfdf7155d313a2806fe95840e7a1d877d8bf41
* Result: SUCCEEDED
* [Build Logs](https://jsv9u8gote.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=90d21a6f-c8d3-4f87-9f38-b4622e07a861%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by malachi-constant** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubDistributedCodeBuild6-jWcl5DLmvupS
* Commit ID: eb8e42d2be05ff583b2523cd9939af453e025bec
* Result: SUCCEEDED
* [Build Logs](https://p0rshq9q5i.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=1432ce21-fccf-4661-8ebd-4e0afd524599%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by malachi-constant** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubLoadTests5656BB24-s6u9F3qN9oFy
* Commit ID: eb8e42d2be05ff583b2523cd9939af453e025bec
* Result: SUCCEEDED
* [Build Logs](https://pciyom2pbh.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=3fdba62e-9adb-429e-8609-96ddc0587ae8%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by malachi-constant** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubStandardCodeBuild8C06-llutOAimTATs
* Commit ID: eb8e42d2be05ff583b2523cd9939af453e025bec
* Result: SUCCEEDED
* [Build Logs](https://jsv9u8gote.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=81decd3e-2ab4-4377-84cc-021e8de2e164%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Pull request merged**